replicatedhq/ship
# Add kustomize build functionality, build by default

What I Did
------------

Update kustomize lifeycle step to build if `dest` is set. The default raw/helm lifecycles now build by default with a dest of `rendered.yaml`. Renamed some fields that could cause breaking changes in ship.yaml apps, but usage of kustomize in lifecycle is low, and we've applied sane defaults.

How I Did it
------------

- rename some fields on the `Kustomize` lifecycle step. `overlay` is now the path to the overlay we create, defaulting to `overlays/ship`. This field used to be called `dest`,
-`dest` is now where the `kustomize build` output will get written. `dest` is optional. If unset we will not do a `kustomize build`.
- If `Dest` is set, we try to get a `render` step for the lifecycle, and use its `Root` property to prefix the `kustomize build` destination (this has no effect in ship init, but may affect using a kustomize step in ship.yaml / replicated.app)
- adjust default lifecycle to build to `rendered.yaml`
- update instructions to use `kubectl apply -f rendered.yaml` instead of the previous `kustomize build overlays/ship | kubectl apply -f -` command.

- update integration tests to check for rendered.yaml
- add `make clean-integration` to clean all test runs
- add support for `SHIP_INTEGRATION_SKIP_CLEANUP_ALL` in integration tests
- change the stable-mysql to use a chart that has deterministic output (`stable/factorio`). Almost every database chart generates a random password, but this one hardcodes a placeholder.
-

Other stuff
---------

- fix a nil pointer panic in replicatedapp/resolver.go
- fix a file handle leak when unpacking github files
- dont include Chart.yaml or values.yaml in `kustomization.yaml#resources`. This is no longer needed for the test I was fixing, but feels like a good idea regardless.

How to verify it
------------

```
ship init ...
diff <(cat rendered.yaml) <(kustomize build overlays/ship)
```

Diff should be empty.

Description for the Changelog
------------

- By default, `ship init` now runs the equivalent of `kustomize build` during execution. Instructions have been updated to suggest a `kubectl apply` instead of the previous `kustomize build ... | kubectl apply -f -` command.
- Fix a file handle leak when unpacking artifacts from github repos

![](https://upload.wikimedia.org/wikipedia/commons/thumb/f/f0/Shipwreck_on_a_rocky_coast.jpg/1920px-Shipwreck_on_a_rocky_coast.jpg)
